packages/mui-base/src/useAutocomplete/useAutocomplete.js

```diff
@@ -300,22 +300,23 @@ export function useAutocomplete(props) {
     let nextFocus = index;
 
     while (true) {
-      // Out of range
-      if (
-        (direction === 'next' && nextFocus === filteredOptions.length) ||
-        (direction === 'previous' && nextFocus === -1)
-      ) {
-        return -1;
-      }
-
       const option = listboxRef.current.querySelector(`[data-option-index="${nextFocus}"]`);
 
-      // Same logic as MenuList.js
       const nextFocusDisabled = disabledItemsFocusable
         ? false
         : !option || option.disabled || option.getAttribute('aria-disabled') === 'true';
 
-      if ((option && !option.hasAttribute('tabindex')) || nextFocusDisabled) {
+      if (direction === 'next' && nextFocus === filteredOptions.length) {
+        if (!nextFocusDisabled) {
+          return -1;
+        }
+        nextFocus = 0;
```

Comment:
Hi @adamhylander , 
i am curios to this issue, so forgive me if i comment to your pr without asking you. 

i just wonder why you not return nextFocus immediately like this
![image](https://github.com/mui/material-ui/assets/9917124/2f8c58d5-2f86-4504-83f9-fbb139f4bc7b)
  this way maybe can help save 1 times of loop "while"

Reply:
Hey @ttlpta,

The reason we set nextFocus = 0/nextFocus = filteredOptions.length is because we wish to run the while loop until an option that is not disabled is found. We essentially wrap the focus around to the end/beginning of the list, depending on direction, and continue to look for an option that isn't disabled. I hope this helps. Let me know if you have any more questions.

Reply:
Hi @adamhylander 
Thank you for ur answer,

I still dont get your point about we need to run the while loop until an option that is not disabled is found.
Because as your condition, you have already check the direction and the next index of item should be focus. I mean the condition `if (direction === 'next' && nextFocus === filteredOptions.length)` and `else if (direction === 'previous' && nextFocus === -1)` is specify to case the disable option is on first/last position of the list. So the next focus index only should be 0, or length of the list.
Btw, Can you send me a test case will be cause of your concern? I have tried some case, but it still works fine when we return imd.

Reply:
Hey again @ttlpta 

Here is index.tsx from the playground environment I used to test the code if you wanna try your suggestion out yourself:
```
import * as React from 'react';
import { StyledEngineProvider } from '@mui/material/styles';
import Autocomplete from '@mui/material/Autocomplete'
import TextField from '@mui/material/TextField'
import Stack from '@mui/material/Stack'

export default function Playground() {
  return (
    <StyledEngineProvider injectFirst>
      <Stack spacing={5}>
        <Autocomplete
          id="disabled-options-demo"
          options={[1, 2, 3, 4, 5, 6, 7]}
          getOptionDisabled={(option) => option % 2 !== 0}
          sx={{ width: 300 }}
          renderInput={(params) => (
            <TextField {...params} label="Disabled options" />
          )}
        />
        <Autocomplete
          id="all-enabled"
          options={[1, 2, 3, 4, 5, 6, 7]}
          sx={{ width: 300 }}
          renderInput={(params) => <TextField {...params} label="All enabled" />}
        />
      </Stack>
    </StyledEngineProvider>
  );
}
```

The if statements `if (direction === 'next' && nextFocus === filteredOptions.length)` and `if (direction === 'previous' && nextFocus === -1)` are not used to specify the case of when the disabled option is on first/last position, it is only used to specify that we are at the beginning or end of the list of options and that we wanna move "out of bounds" (i.e. being at filteredOptions.length[0] and pressing the up key, or being at filteredOptions.length[filteredOptions.length] and pressing the down key).

I would be happy to share my thougts/reasoning on Discord as well if you'd like, Nonetheless, I hope my code is a bit clearer now.

Reply:
Hi @adamhylander 

Perfect, you are correct, i am too focus on the case the first/ last option has been disabled. 

Thank for you answer.

Reply:
Hey @ttlpta

Great! Understanding the logic of nextFocus was confusing for me too. I was wondering what I need to do to pass the last tests? Can you help me with this?